tests/test_dynamics.py

```diff
@@ -194,6 +291,241 @@ def test_torque_derivative_driven(with_contact, with_external_force, cx):
+        Dynamics(
+            DynamicsFcn.TORQUE_DERIVATIVE_DRIVEN, with_contact=with_contact, implicit_soft_contacts=implicit_contact
+        ),
+        False,
+    )
+
+    # Prepare the dynamics
+    ConfigureProblem.initialize(ocp, nlp)
+
+    # Test the results
+    np.random.seed(42)
+    states = np.random.rand(nlp.states.shape, nlp.ns)
+    controls = np.random.rand(nlp.controls.shape, nlp.ns)
+    params = np.random.rand(nlp.parameters.shape, nlp.ns)
+    x_out = np.array(nlp.dynamics_func(states, controls, params))
+
+    if with_contact:
+        contact_out = np.array(nlp.contact_forces_func(states, controls, params))
+        np.testing.assert_almost_equal(
+            x_out[:, 0],
+            [
+                0.6118529,
+                0.785176,
+                0.6075449,
+                0.8083973,
+                -0.3214905,
+                -0.1912131,
+                0.6507164,
+                -0.2359716,
+                0.3886773,
+                0.5426961,
+                0.7722448,
+                0.7290072,
+            ],
+        )
+
+        np.testing.assert_almost_equal(contact_out[:, 0], [-2.444071, 128.8816865, 2.7245124])
+
+    else:
+        np.testing.assert_almost_equal(
+            x_out[:, 0],
+            [
+                0.6118529,
+                0.785176,
+                0.6075449,
+                0.8083973,
+                -0.3024137,
+                -10.3850379,
+                1.6044517,
+                35.8023864,
+                0.3886773,
+                0.5426961,
+                0.7722448,
+                0.7290072,
+            ],
+        )
+
+
+@pytest.mark.parametrize(
+    "dynamics",
+    [DynamicsFcn.TORQUE_ACTIVATIONS_DRIVEN, DynamicsFcn.MUSCLE_DRIVEN],
+)
+def test_implicit_soft_contacts_errors(dynamics):
+    # Prepare the program
+    nlp = NonLinearProgram()
+    nlp.model = biorbd.Model(
+        TestUtils.bioptim_folder() + "/examples/getting_started/models/2segments_4dof_2contacts.bioMod"
+    )
+    nlp.ns = 5
+    nlp.cx = MX
+
+    nlp.u_bounds = np.zeros((nlp.model.nbQ() * 4, 1))
+    ocp = OptimalControlProgram(nlp)
+    nlp.control_type = ControlType.CONSTANT
+    NonLinearProgram.add(
+        ocp,
+        "dynamics_type",
+        Dynamics(dynamics, implicit_soft_contacts=True),
+        False,
+    )
+
+    # Prepare the dynamics
+    if dynamics == DynamicsFcn.TORQUE_DERIVATIVE_DRIVEN:
+        with pytest.raises(
+            TypeError,
+            match=re.escape(
+                "torque_activations_driven() got an unexpected keyword argument " "'implicit_soft_contacts'"
+            ),
+        ):
+            ConfigureProblem.initialize(ocp, nlp)
+    elif dynamics == DynamicsFcn.MUSCLE_DRIVEN:
+        with pytest.raises(
+            TypeError,
+            match=re.escape("muscle_driven() got an unexpected keyword argument " "'implicit_soft_contacts'"),
+        ):
+            ConfigureProblem.initialize(ocp, nlp)
```

Comment:
Why? The signature of the function call should include implicit_soft_contacts..

Reply:
I don't understand your point.
I incorporated a test to verify that it doesnt work with these to dynamics.

Reply:
My point is that it should actually work with these dynamics (even though nothing happens, or an actual raise saying this is not possible)

Reply:
I delete the if.

Reply:
This is also because the error is located higher in a function, we get an error because theses dynamics cannot receive implicit_soft_contacts or implicit_dynamics as an argument.

Reply:
This is what I mean, they should be able to receive it (and raise or not an error associated with their capability (or not) to deal with such dynamics). The fact that they can't receive it means that the signature of the function differs which should not happens (this is a programming error only allowed because Python is based on weak declaration)